backend/src/site_config_cache/mod.rs

```diff
@@ -0,0 +1,258 @@
+const CACHE_NAME: &str = "site_config";
+
+#[derive(Debug, Error)]
+pub enum SiteConfigError {
+    #[error("Redis not configured")] 
+    RedisNotConfigured,
+    #[error(transparent)]
+    Redis(#[from] redis::RedisError),
+    #[error("Redis operation timed out")]
+    Timeout,
+    #[error(transparent)]
+    SerdeJson(#[from] serde_json::Error),
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SiteConfig {
+    pub site_id: String,
+    pub blacklisted_ips: Vec<String>,
+    pub enforce_domain: bool,
+    pub domain: Option<String>,
+    pub updated_at: String,
+    pub version: u64,
+}
+
+#[derive(Clone)]
+pub struct SiteConfigCache {
+    cache: Cache<String, Arc<SiteConfig>>, // keyed by site_id
+    negative_cache: Cache<String, ()>, // tracks recently missing site_ids
+    redis: Arc<RwLock<Option<Arc<Redis>>>>,
+    metrics: Option<Arc<MetricsCollector>>,
+}
+
+impl SiteConfigCache {
+    pub async fn new(redis: Option<Arc<Redis>>, metrics: Option<Arc<MetricsCollector>>) -> Result<Self, SiteConfigError> {
+        let cache = Cache::builder()
+            .max_capacity(50_000)
+            .time_to_live(std::time::Duration::from_secs(60 * 60))
+            .build();
+
+        let negative_cache = Cache::builder()
+            .max_capacity(50_000)
+            .time_to_live(std::time::Duration::from_secs(60))
+            .build();
+
+        Ok(Self { cache, negative_cache, redis: Arc::new(RwLock::new(redis)), metrics })
+    }
+
+    pub async fn is_enabled(&self) -> bool { self.redis.read().await.is_some() }
+
+    pub async fn set_redis(&self, redis: Option<Arc<Redis>>) {
+        let mut lock = self.redis.write().await;
+        *lock = redis;
+    }
+
+    pub async fn get_or_fetch(&self, site_id: &str) -> Result<Option<Arc<SiteConfig>>, SiteConfigError> {
+        if self.negative_cache.contains_key(site_id) {
+            if let Some(m) = &self.metrics { m.increment_cache_lookup(CACHE_NAME, "miss"); }
+            return Ok(None);
+        }
+        if let Some(cfg) = self.cache.get(site_id) {
+            if let Some(m) = &self.metrics { m.increment_cache_lookup(CACHE_NAME, "hit"); }
+            debug!(
+                site_id = %site_id,
+                domain = %cfg.domain.as_deref().unwrap_or(""),
+                enforce_domain = %cfg.enforce_domain,
+                blacklist_count = %cfg.blacklisted_ips.len(),
+                version = %cfg.version,
+                "site-config cache hit"
+            );
+            return Ok(Some(cfg));
+        }
+
+        debug!(site_id = %site_id, "site-config cache miss; fetching from Redis");
+        let key = format!("{}{}", CONFIG_KEY_PREFIX, site_id);
+
+        let redis = self.redis.read().await.clone();
+        let Some(redis) = redis else {
+            if let Some(m) = &self.metrics { m.increment_cache_lookup(CACHE_NAME, "miss"); }
+            return Ok(None);
+        };
```

Comment:
Does this lookup if Redis is setup?
If so, it should probably handle that case further out to not disrupt the code flow.